protocol/lavasession/consumer_session_manager.go

```diff
@@ -85,6 +85,26 @@ func (csm *ConsumerSessionManager) UpdateAllProviders(epoch uint64, pairingList
 	return nil
 }
 
+func (csm *ConsumerSessionManager) UpdateMaxCULimit(virtualEpoch, prevVirtualEpoch uint64) {
+	for key, consumerSessionWithProvider := range csm.pairing {
+		if consumerSessionWithProvider == nil {
+			continue
+		}
+
+		consumerSessionWithProvider.MaxComputeUnits = consumerSessionWithProvider.MaxComputeUnits / (prevVirtualEpoch + 1) * (virtualEpoch + 1)
```

Comment:
add a method on ConsumerSessionsWithProvider to write atomically and use it here (first load and then write).